xmtp_mls/src/groups/sync.rs

```diff
@@ -0,0 +1,771 @@
+        let mls_message_in = MlsMessageIn::tls_deserialize_exact(&envelope.data)?;
+
+        let message = match mls_message_in.extract() {
+            MlsMessageInBody::PrivateMessage(message) => Ok(message),
+            other => Err(MessageProcessingError::UnsupportedMessageType(
+                discriminant(&other),
+            )),
+        }?;
+
+        let intent = provider
+            .conn()
+            .find_group_intent_by_payload_hash(sha256(envelope.data.as_slice()));
+        match intent {
+            // Intent with the payload hash matches
+            Ok(Some(intent)) => self.process_own_message(
+                intent,
+                openmls_group,
+                provider,
+                message.into(),
+                envelope.created_ns,
+                allow_epoch_increment,
+            ),
+            Err(err) => Err(MessageProcessingError::Storage(err)),
+            // No matching intent found
+            Ok(None) => self.process_external_message(
+                openmls_group,
+                provider,
+                message,
+                envelope.created_ns,
+                allow_epoch_increment,
+            ),
+        }
+    }
+
+    fn consume_message(
+        &self,
+        envelope: &GroupMessage,
+        openmls_group: &mut OpenMlsGroup,
+    ) -> Result<(), MessageProcessingError> {
+        let msgv1 = match &envelope.version {
+            Some(GroupMessageVersion::V1(value)) => value,
+            _ => return Err(MessageProcessingError::InvalidPayload),
+        };
+
+        self.client.process_for_id(
+            &msgv1.group_id,
+            EntityKind::Group,
+            msgv1.id,
+            |provider| -> Result<(), MessageProcessingError> {
+                self.process_message(openmls_group, &provider, msgv1, true)?;
+                openmls_group.save(provider.key_store())?;
+                Ok(())
+            },
+        )?;
+        Ok(())
+    }
+
+    pub fn process_messages<'a>(
+        &self,
+        messages: Vec<GroupMessage>,
+        conn: &'a DbConnection<'a>,
+    ) -> Result<(), GroupError> {
+        let provider = self.client.mls_provider(conn);
+        let mut openmls_group = self.load_mls_group(&provider)?;
+
+        let receive_errors: Vec<MessageProcessingError> = messages
+            .into_iter()
+            .map(|envelope| -> Result<(), MessageProcessingError> {
+                retry!(
+                    Retry::default(),
+                    (|| self.consume_message(&envelope, &mut openmls_group))
+                )
+            })
+            .filter_map(Result::err)
+            .collect();
+
+        if receive_errors.is_empty() {
+            Ok(())
+        } else {
+            debug!("Message processing errors: {:?}", receive_errors);
+            Err(GroupError::ReceiveErrors(receive_errors))
+        }
+    }
+
+    pub(super) async fn receive<'a>(&self, conn: &'a DbConnection<'a>) -> Result<(), GroupError> {
```

Comment:
I made `receive` take in a `DbConnection` so that we don't have to unnecessarily create multiple connections during syncing. I kinda wonder if this was the proximate source of our lock contention issues.